### PR TITLE
Add redirects management tool and update _redirects file for all localizations

### PR DESCRIPTION
**Pull request by seokho-son**
This PR proposes a new script tool to manage `_redirects`. In addition, update `_redirects` to remove invalid links and to add all redirects for associated localization languages.

- related issue: #42654

When previously provided pages disappear, we've added redirect to related documents. 
These redirects are listed in https://github.com/kubernetes/website/blob/main/static/_redirects, and currently, there are around 400 entries.

I believe there are several issues in maintaining this `_redirects` file.
1) Some redirects lead to incorrect or non-existent links, and these are not being properly managed. 
2) Kubernetes documentation is localized in more than 10 languages, but these redirects hardly address links in localized languages. Managing this has become a significant challenge. Due to the syntax of _redirects (Netlify), wildcard handling is difficult for localization. 
3) Additionally, there are problems like the list not being sorted, and there's no policy on when to stop a particular redirect. (this tool does not address this issue yet.)

Therefore, the proposed Python program verifies the content of the redirects, regenerates them according to localization, and automatically creates the _redirects. 
1) It checks the list of redirects for inaccessible URLs and ensures that contributors only provide accessible addresses. 
2) If redirects for English are added, it automatically generates redirects for other languages, ensuring that future localization contributors don't have to worry about redirects.

For a easier review process, I recorded what is / how to use Video. 
- https://www.youtube.com/watch?v=ThuhFgOFesk

---

Based on this tool, I found following invalid redirects.

- [89] to_url is NOT available: https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/ 	 (from_url: /docs/concepts/service-catalog/)
- [164] to_url is NOT available: https://kubernetes.io/docs/reference/command-line-tools-reference/cloud-controller-manager/ 	 (from_url: /docs/reference/generated/cloud-controller-manager/)
- [182] to_url is NOT available: https://v1-15.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/ 	 (from_url: /docs/reference/generated/kubernetes-api/v1.15/)
- [183] to_url is NOT available: https://v1-16.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/ 	 (from_url: /docs/reference/generated/kubernetes-api/v1.16/)
- [184] to_url is NOT available: https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/ 	 (from_url: /docs/reference/generated/kubernetes-api/v1.17/)
- [185] to_url is NOT available: https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/ 	 (from_url: /docs/reference/generated/kubernetes-api/v1.18/)
- [174] to_url is NOT available: https://kubernetes.io/docs/reference/glossary/approver/ 	 (from_url: /docs/reference/glossary/maintainer/)
- [405] to_url is NOT available: https://kubernetes.io/docs/setup/production-environment/on-premises-vm/cloudstack/ 	 (from_url: /docs/setup/on-premises-vm/cloudstack/)
- [406] to_url is NOT available: https://kubernetes.io/docs/setup/production-environment/on-premises-vm/dcos/ 	 (from_url: /docs/setup/on-premises-vm/dcos/)
- [407] to_url is NOT available: https://kubernetes.io/docs/setup/production-environment/on-premises-vm/ovirt/ 	 (from_url: /docs/setup/on-premises-vm/ovirt/)
- [231] to_url is NOT available: https://kubernetes.io/docs/tasks/administer-cluster/upgrade-downgrade/kubeadm-upgrade-ha/ 	 (from_url: /docs/tasks/administer-cluster/kubeadm-upgrade-ha/)
- [235] to_url is NOT available: https://kubernetes.io/docs/tasks/administer-cluster/memory-available.sh 	 (from_url: /docs/tasks/administer-cluster/out-of-resource/memory-available.sh)
- [276] to_url is NOT available: https://kubernetes.io/docs/tasks/run-application/upgrade-pet-set-to-stateful-set/ 	 (from_url: /docs/tasks/manage-stateful-set/upgrade-pet-set-to-stateful-set/)
- [345] to_url is NOT available: https://kubernetes.io/docs/community 	 (from_url: /events/)
- [372] to_url is NOT available: https://kubernetes.io/blog/2022/02/10/image-registry-redirect/ 	 (from_url: /image-registry-change/)
- [370] to_url is NOT available: https://kubernetes.io/blog/2022/02/10/image-registry-redirect/ 	 (from_url: /image-registry-redirect/)
- [351] to_url is NOT available: https://kubernetes.io/docs/community 	 (from_url: /media/)
- [352] to_url is NOT available: https://kubernetes.io/docs/community 	 (from_url: /news/)
- [10] to_url is NOT available: https://kubernetes.io/no/docs/home/ 	 (from_url: /no/docs/)

And I updated some of them or removed useless redirects. (PTAL the redirects I removed `-----` are actually useless or not)

```diff
-----/docs/concepts/service-catalog/,/docs/concepts/extend-kubernetes/service-catalog/,301
-----/docs/reference/generated/cloud-controller-manager/,/docs/reference/command-line-tools-reference/cloud-controller-manager/,301
-----/docs/reference/generated/kubernetes-api/v1.15/,https://v1-15.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/,301
-----/docs/reference/generated/kubernetes-api/v1.16/,https://v1-16.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/,301
-----/docs/reference/generated/kubernetes-api/v1.17/,https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/,301
-----/docs/reference/generated/kubernetes-api/v1.18/,https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/,301
/docs/reference/glossary/maintainer/,/docs/reference/glossary/?community=true#term-approver,301
-----/docs/setup/on-premises-vm/cloudstack/,/docs/setup/production-environment/on-premises-vm/cloudstack/,301
-----/docs/setup/on-premises-vm/dcos/,/docs/setup/production-environment/on-premises-vm/dcos/,301
-----/docs/setup/on-premises-vm/ovirt/,/docs/setup/production-environment/on-premises-vm/ovirt/,301
/docs/tasks/administer-cluster/kubeadm-upgrade-ha/,/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/,301
/docs/tasks/administer-cluster/out-of-resource/memory-available.sh,/docs/tasks/administer-cluster/manage-resources/memory-constraint-namespace/,301
-----/docs/tasks/manage-stateful-set/upgrade-pet-set-to-stateful-set/,/docs/tasks/run-application/upgrade-pet-set-to-stateful-set/,301
/events/,/community,301
/image-registry-change/,/blog/2023/03/10/image-registry-redirect/,302
/image-registry-redirect/,/blog/2023/03/10/image-registry-redirect/,302
/media/,/community,301
/news/,/community,301
-----/no/docs/,/no/docs/home/,301!
```

### CODE REVIEW

**Comment by seokho-son** [comment]
I hope to receive reviews from @kubernetes/sig-docs-leads. I believe we need a consensus for this kind of approach. (how to manage the redirects)

**Comment by sftim** [comment]
As a tech lead for SIG Docs, I'd take a steer from the localization subproject. What does the localization subproject conclude about these proposed changes?

File: static/gen-redirects.csv
Comment (sftim) [inline static/gen-redirects.csv:1]: I would put this file somewhere inside `data` and not in `static`. No need to publish it as https://k8s.io/gen-redirects.csv (in my opinion).

**Review by sftim — COMMENTED** [review]
If we want to generate `_redirects`, maybe https://honnef.co/articles/generating-netlify-_redirects-from-hugo/ suggests a starting point for a possible approach?

**Comment by tengqm** [comment]
This doesn't look to me like a viable/sustainable approach for managing redirects.

If you have tried the current `scripts/linkchecker.py` script, you will see that
we are able to run link checks along with redirect checks using a simpler method.
That tool is not perfect, lacking support to check blog paths or support to parse
examples path or static redirects (such as '/dockershim'). However, it is a good
starting point for further improvements. I am not a big fan of adding a lot of
tools overlapping with each other.

Speaking of redirects for localized contents, I had a proposal 4 years ago (https://github.com/kubernetes/website/pull/18374). It is a much simpler approach
to make sure that we (the 10+ localization teams) don't need to prefix each and
every link target with a language code. The `ref` and `relref` shortcode are there
just for this purpose. The proposal wasn't quite received by the reviewers, I believe.
It looks like a huge churn to the whole site. However, it is a solution that,
once applied, will benefit the whole community in the long run.

**Comment by sftim** [comment]
@seokho-son, would you be willing to open an issue for this? We might also want to make a GitHub Discussion to run through the different options.

**Comment by seokho-son** [comment]
> @seokho-son, would you be willing to open an issue for this? We might also want to make a GitHub Discussion to run through the different options.

Hi @sftim I opened an issue for this. :) 
https://github.com/kubernetes/website/issues/42654

I haven't opened a Discussion yet. I think we can still use #42654 to explore other viable options. If we encounter too many diverse options and discussions, I'll open a new Discussion. is it ok?

File: static/gen-redirects.csv
Comment (seokho-son) [reply static/gen-redirects.csv:1]: Sure, let me relocate it after we decide to generate `_redirects` using this tool.

**Comment by seokho-son** [comment]
> 

Hi @tengqm, thanks for sharing your insight! I remember you mentioned the redirects issue several times in multiple PRs.

> If you have tried the current scripts/linkchecker.py script, you will see that
> we are able to run link checks along with redirect checks using a simpler method.
> That tool is not perfect, lacking support to check blog paths or support to parse
> examples path or static redirects (such as '/dockershim'). However, it is a good
> starting point for further improvements. I am not a big fan of adding a lot of
> tools overlapping with each other.

- I'm aware of linkchecker.py, but I believe the purpose of this tool differs slightly. My aim is to ensure that both contributors and maintainers generate a correct _redirects file every time they modify it. By doing so, we can be confident that there are no issues with _redirects and that localization is properly managed. From this perspective, I see it as a sustainable approach, though it's not fully automated, unfortunately.

> Speaking of redirects for localized contents, I had a proposal 4 years ago (https://github.com/kubernetes/website/pull/18374). It is a much simpler approach
> to make sure that we (the 10+ localization teams) don't need to prefix each and
> every link target with a language code. The ref and relref shortcode are there
> just for this purpose. The proposal wasn't quite received by the reviewers, I believe.
> It looks like a huge churn to the whole site. However, it is a solution that,
> once applied, will benefit the whole community in the long run.

- If we have a simpler automated way like the option you suggested, I'd love to see it implemented. But, I guess changing huge churn to the whole site is also a challenge. Considering it is just to handle only the redirects.

**Comment by tengqm** [comment]
I'm not **yet** convinced that the `_redirects` can be auto-generated.
Even if there is a slight possibility, I'm worrying that it will introduce more subtle bugs than benefits.

**Comment by tengqm** [comment]
Well, there do exist a good opportunity to check if there are invalid/dangling redirect entries, and that could be a simpler tool.

**Comment by seokho-son** [comment]
> I'm not **yet** convinced that the `_redirects` can be auto-generated. Even if there is a slight possibility, I'm worrying that it will introduce more subtle bugs than benefits.

@tengqm Sorry for delayed response. (too many works ;) )

I'd like to clarify your comment:

> I'm not yet convinced that the _redirects can be auto-generated.

Are you suggesting that _redirects is difficult to auto-generate regardless of the tool introduced in this PR? 
Or are you implying that the tool proposed might not generate appropriate _redirects?

For your information, the [_redirects](https://github.com/kubernetes/website/blob/329d0f94996b6365eca2fd1607d87fdf20afbae9/static/_redirects) file in this PR was solely generated by the proposed tool without any manual adjustments. I've tested its output using the [Redirects Playground](https://play.netlify.com/redirects) https://play.netlify.com/redirects, and it validated successfully.

FYI: the [_redirects](https://github.com/kubernetes/website/blob/329d0f94996b6365eca2fd1607d87fdf20afbae9/static/_redirects) file is in this PR was generated (and not touched manually) by the proposed tool. 
Tested the result with https://play.netlify.com/redirects Redirects Playground. 
And it says validation is success. 

> Even if there is a slight possibility, I'm worrying that it will introduce more subtle bugs than benefits.

Do you expect an alternative suggestion? Or are you hoping for improvements to the current tool? (or just need verification?)

**Comment by tengqm** [comment]
I'm afraid that the `_redirects` file is an expression of intent rather than something you can summarize from existing data. For example, when we are renaming a file, say 'topology-aware-hints.md', to 'topology-aware-routing.md', this renaming is an intent. We add an entry to the `_redirects` file so that anyone who had bookmarked the previous URL can still access the new file rather than receiving a 404. This change (and the intent) cannot be magically captured by any automation file. Yes, we can create a change record for this somewhere and then having the automation tool to regenerate the `_redirects` file. Do you believe this indirection is bringing us any benefit? Isn't it good enough to just add an entry to the `_redirects` file instead?

There are some more complicated scenarios. Some localization teams may want to add a redirect entry for a particular page, some teams may believe it is not necessary. Do we have a generic solution to this?

I'm not in favor of any over-engineering effort that brings in little to no advantages, but turns a simple issue to a complex one. It looks like we solved a problem. But, we didn't. We are creating a new asset that is, when broken, much more difficult to resolve than the original one.

**Comment by seokho-son** [comment]
@tengqm Thanks for the comment. Let me share my opinion on your comment. :)

> I'm afraid that the _redirects file is an expression of intent rather than something you can summarize from existing data. For example, when we are renaming a file, say 'topology-aware-hints.md', to 'topology-aware-routing.md', this renaming is an intent. We add an entry to the _redirects file so that anyone who had bookmarked the previous URL can still access the new file rather than receiving a 404. This change (and the intent) cannot be magically captured by any automation file. 

- I agree. (and I assume you know the intention of this tool is not to generate redirects by automatically guessing those intents)

> Yes, we can create a change record for this somewhere and then having the automation tool to regenerate the _redirects file. Do you believe this indirection is bringing us any benefit? Isn't it good enough to just add an entry to the _redirects file instead?

- Based on the current status of contributions and management of redirects, I believe this indirection is bringing us benefit especially for localization aspects. 
- Localization contributors (or main maintaining contributors) seem unable to actively respond to redirect changes for English pages. If a redirect is added to an English page, and there are (several) localized documents corresponding to that English page, redirects are mostly required for those localized pages as well. 
If not addressed, 404 errors occur in the localized pages. (Once the localization team realizes the 404, they typically add redirects for the localized pages then, or they overlook it.)
- Moreover, when a contributor adds (or modifies) a redirect for an English page, they tend to not include redirects for all related localized pages together (the contributor is hard to check which localized pages are related).
- Furthermore, when a related localized document is newly added in situations where there is an existing redirect to the English page, the localization contributor tends to not add the necessary redirect (for this part, we can guide localization contributors). Anyway, Localization efforts usually face challenges because the number of contributors for non-English content is generally smaller than for English content. We'd better reduce efforts on maintaining localizations.

- As @sftim mentioned above, for localization, there isn't active discussion or guidance on handling redirects. 
(There's a need to discuss solutions // The fact that the discussion hasn't taken place yet is my guilty. ;) )
Anyway, I'd say the tool will at least help localization contributors not to worry about redirect changes on English pages.

> There are some more complicated scenarios. Some localization teams may want to add a redirect entry for a particular page, some teams may believe it is not necessary. Do we have a generic solution to this?
- This tool does not provide a generic solution for the scenarios (yet). 
- However, up until now, I haven't found the scenario (Some localization teams may want to add a redirect entry for a particular page) in the existing _redirects file, except for a few special cases (/pt/* /pt-br/:splat, /zh/* /zh-cn/:splat). I assume that localization teams haven't been using _redirects solely for redirecting localized pages.
- Anyway, If we need, this tool can provide a section for exceptions.

> I'm not in favor of any over-engineering effort that brings in little to no advantages, but turns a simple issue to a complex one. It looks like we solved a problem. But, we didn't. We are creating a new asset that is, when broken, much more difficult to resolve than the original one. 
- Sure, we don't need over-engineering. :) I am not saying that this tool should be included. We need to take a better way if we have.
- I'm uncertain about the potential breaking points you mentioned. The indirect file feels intuitive (at least to me), since it is similar with the original _redirect. I don't perceive it as complicated. However, the source code of this tool itself might appear intricate, possibly because I'm not a seasoned developer. Any guidance on this would be greatly appreciated.

**Comment by tengqm** [comment]
> If a redirect is added to an English page, and there are (several) localized documents corresponding to that English page, redirects are mostly required for those localized pages as well.

Suppose we have content/en/docs/foo/page1.md relocated to content/en/docs/bar/page2.md, and a redirect entry is added to `_redirects` file for this change. How does this change impact the `content/<lang>/foo/page1.md` ? I don't get it. Aren't we using localized page for links?

> when a contributor adds (or modifies) a redirect for an English page, they tend to not include redirects for all related localized pages together (the contributor is hard to check which localized pages are related).

Then, what are we supposed to do in this case? Add redirects for all localizations? For each and every English redirect records, we add 10+ more redirect records for all other languages?

> when a related localized document is newly added in situations where there is an existing redirect to the English page, the localization contributor tends to not add the necessary redirect ...

Adding redirects for all localizations is a terrible idea. We could have saved that huge effort since Hugo already provided `ref` and `relref` shortcode for this. Why don't we use the builtin mechanism? Maintaining a huge, and ever-growing `_redirects` file is not fun. Please consider the long term impact.

Why don't we explore the "native" multilingual support from Hugo? I don't think changing the English upstream to use `ref` or `relref` is that hard. It will save a huge amount of time for all localization teams. The sad thing is that this approach is not well understood by some approvers.

**Comment by sftim** [comment]
> I don't think changing the English upstream to use `ref` or `relref` is that hard.

We can do that, but even if we do, we _still_ need to manage redirects and to make that work across different localizations. For example, we redirect some removed pages to the closest still-relevant page. We also have shortlinks like https://k8s.io/dockerfile and (if it existed) https://k8s.io/uk/dockerfile

**Comment by tengqm** [comment]
To clarify, if we switch to `ref` or `relref` directives, we will be using the relative path to the markdown files as links. There won't be a huge burden to manage redirects. A link either works or it doesn't. Once we fix the link in the English page, we don't have to fix it for any localization teams.

**Comment by seokho-son** [comment]
It seems that this PR has not achieved further consensus, and it looks like it's time to close it. I hope the tooling example from this PR has been helpful in broadening our perspectives. Thanks for the discussion. @tengqm @sftim  I think further discussion can be made in other PRs or issues. I will be utilizing the script personally to update static/_redirects. :) 

/close

**Pull request closed without merging**